PopIII Simulation

psharda opened the pull request:
This PR will add a PopIII simulation to quokka. It will only test that the simulation starts successfully and runs for a few timesteps - it's not meant to be a _production_ simulation.

Commit:
added PopIII setup

Commit:
added rotation

Commit:
added minimum density below which chemistry should not be performed

Commit:
final touches

psharda committed:
Merge remote-tracking branch 'origin/development' into popiii

Commit:
update with popiii

Commit:
updated

psharda committed:
bug fix

psharda committed:
decrease timesteps

Commit:
use numdens

Commit:
verbose

psharda committed:
fix bug

psharda committed:
Merge branch 'popiii' of github.com:psharda/quokka into popiii

psharda committed:
define proper values

psharda committed:
fix bug

Commit:
try

psharda committed:
fixed bug

psharda committed:
try2

psharda committed:
reset density floor

psharda committed:
define temp outside parallelFor

psharda committed:
Merge branch 'quokka-astro:development' into popiii

psharda committed:
use BL_Profile

psharda committed:
inputs for perturbations.py direcrly read from the .in file

psharda committed:
Merge branch 'quokka-astro:development' into popiii

psharda committed:
updated cmakelists

psharda commented:
/azp run

psharda committed:
correct includes order

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

psharda committed:
add {} to variable declaration

psharda committed:
added author name

psharda committed:
remove duplicate include

BenWibking commented on src/PopIII/TurbDataReader.cpp:
```diff
@@ -0,0 +1,119 @@
+//==============================================================================
```
Can we re-use the TurbDataReader from the other problems? Maybe move it to `src/` and not in the problem-specific directories?

psharda replied:
@BenWibking TurbDataReader has both .cpp and .hpp files. Do I move them both to `src/`? If yes, then while adding a test in, say, `StarCluster/CMakeLists.txt`, do I compile both .cpp and .hpp?

BenWibking replied:
Yes. You can compile the .cpp file. The .hpp file is just `#include`-ed.

BenWibking commented on src/PopIII/perturbation.py:
```diff
@@ -0,0 +1,311 @@
+import h5py
```
Can we put this in `scripts/` and re-use it for multiple problems?

BenWibking commented on src/CMakeLists.txt:
```diff
@@ -154,4 +157,3 @@ add_subdirectory(ShockCloud)
 add_subdirectory(PassiveScalar)
 add_subdirectory(FCQuantities)
 add_subdirectory(SphericalCollapse)
-add_subdirectory(NSCBC)
```
Is this change a typo? I think we want to keep this test problem.

psharda replied:
oops. yeah this is a typo

BenWibking commented on src/RadhydroSimulation.hpp:
```diff
@@ -360,6 +362,7 @@ template <typename problem_t> void RadhydroSimulation<problem_t>::readParmParse(
 		amrex::ParmParse hpp("primordial_chem");
 		hpp.query("enabled", enableChemistry_);
 		hpp.query("max_density_allowed", max_density_allowed);
+		hpp.query("min_density_allowed", min_density_allowed);
```
maybe add this comment here as well: `// dont do chemistry in cells with densities below the minimum density specified`

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,431 @@
+};
+
+template <> struct HydroSystem_Traits<PopIII> {
+	static constexpr bool reconstruct_eint = false;
```
If multiphase gas develops, this should be set to `true`.

psharda replied:
woah, ofcourse we have multiphase gas! Didn't know this

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,431 @@
+
+	auto const &state = state_new_cc_[lev].const_arrays();
+	auto tag = tags.arrays();
+	amrex::Print() << "start error est " << std::endl;
```
I don't think we want to commit this print statement this unless there's a way to turn it on/off.

psharda replied:
ah lol, yes.

psharda replied:
will do it at the end before this PR is merged

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,431 @@
+template <> struct Physics_Traits<PopIII> {
+	// cell-centred
+	static constexpr bool is_hydro_enabled = true;
+	static constexpr bool is_chemistry_enabled = false;	     // in the future, this could point to microphysics, and set to true
```
We can get rid of this parameter, right? Everything is set with `numMassScalars`, plus the runtime parameters in the `.in` file, no?

psharda replied:
Yeah, we don't really use `is_chemistry_enabled`

BenWibking replied:
We should probably make a separate PR to get rid of this.

psharda replied:
True.

psharda committed:
include NSCBC

psharda committed:
added comments

psharda committed:
set reconstruct_eint to true

psharda committed:
move perturbations to src/

psharda committed:
moved perturbatinos to src/

psharda committed:
move perturb to src

psharda committed:
moved perturb to src

psharda committed:
fix bug

psharda committed:
re-add enabled

psharda commented:
/azp run

psharda committed:
updated params to match PopIII expectations

psharda committed:
remove turbdatareader

psharda committed:
remove turbdatareader

psharda committed:
move turbdatareader to src

psharda committed:
dont refine outside core

psharda committed:
Merge remote-tracking branch 'origin/development' into popiii

psharda committed:
remove is chem enabled

psharda committed:
Merge branch 'development' into popiii

psharda committed:
define redshift

psharda committed:
reset params

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'quokka-astro:development' into popiii

psharda committed:
Merge branch 'quokka-astro:development' into popiii

psharda committed:
modify

Commit:
increase density floor

Commit:
return invalid state if mass scalars have negative partial densities

Commit:
enforce limits on mass scalars

BenWibking committed:
add LLF fallback solver

Commit:
use if constexpr

Commit:
Merge remote-tracking branch 'origin/specie-limits' into popiii

Commit:
Merge remote-tracking branch 'ben/BenWibking/llf-fallback' into popiii

BenWibking committed:
fix typo

Commit:
Merge remote-tracking branch 'ben/BenWibking/llf-fallback' into popiii

Commit:
Merge remote-tracking branch 'origin/development' into popiii

Commit:
set temp floor based on redshift

psharda committed:
Merge branch 'quokka-astro:development' into specie-limits

Commit:
print cell coords

Commit:
print cell coords

Commit:
check for dimensions

Commit:
amrex spelling

Commit:
new function for printing coords

Commit:
fix error

Commit:
declare const

Commit:
fix bug

Commit:
Merge remote-tracking branch 'origin/specie-limits' into popiii

Commit:
remove clangtidy suggestion

Commit:
drive turb in entire box

Commit:
Merge remote-tracking branch 'origin/specie-limits' into popiii

BenWibking committed:
use minmod limiting for PLM

BenWibking committed:
write multifab to disk if hydro update fails

BenWibking committed:
call MPI_Abort

Commit:
Merge remote-tracking branch 'ben/BenWibking/use-minmod' into popiii

Commit:
Merge remote-tracking branch 'ben/BenWibking/output-multifab-on-crash' into popiii

Commit:
debuggin

Commit:
debugging

psharda committed:
Merge branch 'development' into popiii

Commit:
dont do chem if dt < 0

Commit:
use speed ceiling

Commit:
ensure pressure eqbm in ICs

Commit:
checks

Commit:
Merge branch 'popiii' of github.com:psharda/quokka into popiii

Commit:
add velx and pressure as derived vars

Commit:
add exmaple of refinement restricted to certain  cell positions to avoid refinement at cloud edge

Commit:
abort if rho = 0;

Commit:
check hydro state in more locations

Commit:
fix tab

Commit:
reduce validstates boolean

Commit:
add function call name keyword

Commit:
enforceLimits returns a bool

Commit:
enforceLimits returns a bool

Commit:
write pltfile when rho = 0 encountered and abort

Commit:
indent

Commit:
added dummy test

Commit:
fix bug

Commit:
updated

Commit:
reset refinement

Commit:
added sound speed

Commit:
added sound speed

psharda committed:
do not abort

psharda committed:
refine on 64 cells

psharda committed:
update amrex

psharda committed:
Merge branch 'development' into popiii

psharda committed:
delete copiii

psharda committed:
remove debugs

psharda committed:
remove debugs

psharda committed:
remove debugs

psharda committed:
updated

psharda committed:
remove speed ceiling

Commit:
renormalize species post flooring

psharda committed:
renormalize mass scalars in enforce limits

psharda committed:
resolve spell check

psharda committed:
Merge branch 'development' into popiii

psharda committed:
dont need velocity large anymore in states valid

psharda committed:
undo some stuff

psharda committed:
remove debug printf

psharda committed:
remove debug printf

psharda committed:
remove debug printf

psharda committed:
remove debug printf

psharda committed:
revert enforceLimits as boolean

psharda committed:
revert enforceLimits boolean

psharda committed:
fix spaces

psharda committed:
sample input file for main repo

psharda commented:
@BenWibking how do you check what causes clang-tidy to fail? It doesn't show a useful error message that I can debug

BenWibking commented:
> @BenWibking how do you check what causes clang-tidy to fail? It doesn't show a useful error message that I can debug

~~It does. Somehow the command `exit 1` got added to the action, which as you can imagine, causes it to always fail: https://github.com/quokka-astro/quokka/actions/runs/6327998862/job/17185136213?pr=368#step:6:1~~

Edit: ah ok, so what it's doing is this: https://github.com/quokka-astro/quokka/blob/a37934f74c6de738321d6898906afd4a054f26bd/.github/workflows/clang-tidy.yml#L32

So if it produces any warnings, it causes it to fail. We can probably remove this from the action, since resolving the comments is required for a merge anyway.

psharda commented:
Hmm, `HydroShocktubeCMA` does not like the renormalization of mass scalars in `EnforceLimits`. Seems that at some point `sp_sum = 0` here: https://github.com/psharda/quokka/blob/7bf9c792d89477a15971d4470b84335d04a1379a/src/hydro_system.hpp#L764

BenWibking commented:
> Hmm, `HydroShocktubeCMA` does not like the renormalization of mass scalars in `EnforceLimits`. Seems that at some point `sp_sum = 0` here: https://github.com/psharda/quokka/blob/7bf9c792d89477a15971d4470b84335d04a1379a/src/hydro_system.hpp#L764

This is getting rather complicated. Maybe move this into its own PR and we can debug it there?

Commit:
renormalize mass scalars post enforcing limit

psharda commented:
> > Hmm, `HydroShocktubeCMA` does not like the renormalization of mass scalars in `EnforceLimits`. Seems that at some point `sp_sum = 0` here: https://github.com/psharda/quokka/blob/7bf9c792d89477a15971d4470b84335d04a1379a/src/hydro_system.hpp#L764
> 
> This is getting rather complicated. Maybe move this into its own PR and we can debug it there?

#405

Commit:
only normalize if sum > 0

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

psharda committed:
updated

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

psharda committed:
update

psharda committed:
merged spsum

psharda commented:
/azp run

psharda commented:
@BenWibking all tests pass but the clang-tidy one fails

psharda committed:
Merge branch 'development' into popiii

psharda committed:
cosmetic

BenWibking commented:
> @BenWibking all tests pass but the clang-tidy one fails

This indicates you have an unaddressed clang-tidy comment. You can either apply the fix that clang-tidy wants, or you can add `// NOLINT` to suppress the warning.

psharda commented:
> > @BenWibking all tests pass but the clang-tidy one fails
> 
> This indicates you have an unaddressed clang-tidy comment. You can either apply the fix that clang-tidy wants, or you can add `// NOLINT` to suppress the warning.

Where can I find the unaddressed comment

BenWibking commented:
> > > @BenWibking all tests pass but the clang-tidy one fails
> > 
> > 
> > This indicates you have an unaddressed clang-tidy comment. You can either apply the fix that clang-tidy wants, or you can add `// NOLINT` to suppress the warning.
> 
> Where can I find the unaddressed comment

It looks like you resolved a few clang-tidy comments by clicking 'resolve conversation' without applying any code changes. I would look at those. It should also be available in the GitHub action logs for the clang-tidy-review action.

BenWibking commented:
For example, for the most recent commit to PR #406, you can see the clang-tidy warnings in this logfile: https://github.com/quokka-astro/quokka/actions/runs/6436937742/job/17481260240#step:4:1084

BenWibking commented:
You can see the warnings here: https://github.com/quokka-astro/quokka/actions/runs/6437054345/job/17481553174?pr=368#step:4:1085

psharda committed:
remove actual_eos_data.H

psharda committed:
resolve clang tidy comments

psharda commented:
@BenWibking clang-tidy is likely failing because of this clang-diagnostic-error: https://github.com/quokka-astro/quokka/actions/runs/6444002563/job/17501485519#step:4:1534

But I don't understand why it doesn't find the `actual_eos_data.H` file. All the other tests can find this file (otherwise they would also throw an error right).

BenWibking commented:
> @BenWibking clang-tidy is likely failing because of this clang-diagnostic-error: https://github.com/quokka-astro/quokka/actions/runs/6444002563/job/17501485519#step:4:1534
> 
> But I don't understand why it doesn't find the `actual_eos_data.H` file. All the other tests can find this file (otherwise they would also throw an error right).

Yeah, I agree with this. I'm also not sure why it's doing this, since CMake is supposed to pass all of the include directories to clang-tidy (e.g., https://stackoverflow.com/questions/39455090/clang-tidy-cant-find-my-header-files). I see lots of Stackoverflow posts about this :/

BenWibking commented:
However, Microphysics and Castro also use clang-tidy now. So we might be able to copy their settings or GitHub actions.

psharda commented:
> However, Microphysics and Castro also use clang-tidy now. So we might be able to copy their settings or GitHub actions.

I don't see a clang-tidy.yml file in either Castro or Microphysics. But they don't use CMake anyway right (except the part we wrote for Microphysics)

BenWibking commented:
> > However, Microphysics and Castro also use clang-tidy now. So we might be able to copy their settings or GitHub actions.
> 
> I don't see a clang-tidy.yml file in either Castro or Microphysics. But they don't use CMake anyway right (except the part we wrote for Microphysics)

Right. Hmm, I must have been thinking of what AMReX does, which is not quite the same as what we do. You can see what AMReX does here: https://github.com/AMReX-Codes/amrex/pull/3176 and here: https://github.com/AMReX-Codes/amrex/blob/development/.github/workflows/clang.yml

psharda commented:
I'm just thinking - shouldn't this error also appear for the Primordial Chem test? Because it also includes the actual_eos_data.H file. Why does it only show up for PopIII

BenWibking commented:
> I'm just thinking - shouldn't this error also appear for the Primordial Chem test? Because it also includes the actual_eos_data.H file. Why does it only show up for PopIII

I don't know 🤷‍♂️

Commit:
updated microp

psharda commented:
I updated microphysics but now homebrew seems to barf

BenWibking commented:
> I updated microphysics but now homebrew seems to barf

It's a recurring problem with Homebrew. You can google this and find a lot of github/stackoverflow issues :/ It will probably be fixed in a day or two.

psharda commented:
haha okay. Feels like I'm discovering a whole new world of issues that go beyond quokka.

BenWibking commented:
> haha okay. Feels like I'm discovering a whole new world of issues that go beyond quokka.

welcome to code development 🙃 

this is the issue: https://github.com/actions/runner-images/issues/8500

psharda committed:
Merge branch 'quokka-astro:development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda commented:
Homebrew is fixed now but we are back to the clang tidy error on file actual_eos_data.H not found

BenWibking commented:
Can you try running this command on your local computer and see if you get the same error:

```
clang-tidy-16 -p=build --config-file="src/.clang-tidy" -line-filter='[{"name":"src/Chemistry.hpp","lines":[[34,34],[37,37],[68,72],[86,89],[95,98]]},{"name":"src/PopIII/popiii.cpp","lines":[[1,475]]},{"name":"src/PopIII/popiii.hpp","lines":[[1,23]]},{"name":"src/RadhydroSimulation.hpp","lines":[[116,117],[376,377],[503,503]]}]' src/Chemistry.hpp src/PopIII/popiii.cpp src/PopIII/popiii.hpp src/RadhydroSimulation.hpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp --export-fixes=clang_tidy_review.yaml
```

psharda committed:
fix bug

psharda commented:
ah this was very helpful, caught the bug:

```
ps3459@gadi-login-02:/g/data1b/jh2/ps3459/quokka>clang-tidy -p=build --config-file="src/.clang-tidy" -line-filter='[{"name":"src/Chemistry.hpp","lines":[[34,34],[37,37],[68,72],[86,89],[95,98]]},{"name":"src/PopIII/popiii.cpp","lines":[[1,475]]},{"name":"src/PopIII/popiii.hpp","lines":[[1,23]]},{"name":"src/RadhydroSimulation.hpp","lines":[[116,117],[376,377],[503,503]]}]' src/Chemistry.hpp src/PopIII/popiii.cpp src/PopIII/popiii.hpp src/RadhydroSimulation.hpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp --export-fixes=clang_tidy_review.yaml
164538 warnings generated.
345999 warnings and 1 error generated.
Error while processing /g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp.
510724 warnings and 1 error generated.
Error while processing /g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.hpp.
677967 warnings and 1 error generated.
Error while processing /g/data1b/jh2/ps3459/quokka/src/RadhydroSimulation.hpp.
804205 warnings and 1 error generated.
Error while processing /g/data1b/jh2/ps3459/quokka/src/TurbDataReader.cpp.
930443 warnings and 1 error generated.
Error while processing /g/data1b/jh2/ps3459/quokka/src/TurbDataReader.cpp.
1056681 warnings and 1 error generated.
Error while processing /g/data1b/jh2/ps3459/quokka/src/TurbDataReader.cpp.
1182919 warnings and 1 error generated.
Error while processing /g/data1b/jh2/ps3459/quokka/src/TurbDataReader.cpp.
warning: invalid configuration value '0' for option 'modernize-make-shared.IncludeStyle' [clang-tidy-config]
warning: invalid configuration value '0' for option 'modernize-make-unique.IncludeStyle' [clang-tidy-config]
/g/data1b/jh2/ps3459/quokka/src/Chemistry.hpp:37:2: warning: variable 'tiny_profiler_165' of type 'amrex::TinyProfiler' can be declared 'const' [misc-const-correctness]
        BL_PROFILE("Chemistry::computeChemistry()");
        ^
/g/data1b/jh2/ps3459/quokka/extern/amrex/Src/Base/AMReX_BLProfiler.H:487:27: note: expanded from macro 'BL_PROFILE'
#define BL_PROFILE(fname) BL_PROFILE_IMPL(fname, __COUNTER__)
                          ^
/g/data1b/jh2/ps3459/quokka/extern/amrex/Src/Base/AMReX_BLProfiler.H:488:43: note: expanded from macro 'BL_PROFILE_IMPL'
#define BL_PROFILE_IMPL(funame, counter)  amrex::TinyProfiler BL_PROFILE_PASTE(tiny_profiler_, counter)((funame)); \
                                          ^
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:268:11: error: cannot assign to variable 'rhotot' with const-qualified type 'const amrex::Real' (aka 'const double') [clang-diagnostic-error]
                        rhotot += state.xn[n] * spmasses[n]; // spmasses contains the masses of all species, defined in EOS
                        ~~~~~~ ^
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:264:21: note: variable 'rhotot' declared const here
                amrex::Real const rhotot = 0.0;
                ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:357:4: warning: variable 'cs' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
                        ^
                                    const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:377:4: warning: variable 'Eint' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real Eint = state[bx](i, j, k, HydroSystem<PopIII>::internalEnergy_index);
                        ^
                                    const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:392:4: warning: variable 'Pgas' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real Pgas = HydroSystem<PopIII>::ComputePressure(state[bx], i, j, k);
                        ^
                                    const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:418:4: warning: variable 'pressure' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        Real pressure = HydroSystem<PopIII>::ComputePressure(state[bx], i, j, k);
                        ^
                             const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:421:4: warning: variable 'cs' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
                        ^
                                    const
Suppressed 1183244 warnings (1182839 in non-user code, 53 due to line filter, 352 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Found compiler error(s).
```

psharda commented:
ugh, still fails. On gadi now, it produces no errors:

```
ps3459@gadi-login-02:/g/data1b/jh2/ps3459/quokka>clang-tidy -p=build --config-file="src/.clang-tidy" -line-filter='[{"name":"src/Chemistry.hpp","lines":[[34,34],[37,37],[68,72],[86,89],[95,98]]},{"name":"src/PopIII/popiii.cpp","lines":[[1,475]]},{"name":"src/PopIII/popiii.hpp","lines":[[1,23]]},{"name":"src/RadhydroSimulation.hpp","lines":[[116,117],[376,377],[503,503]]}]' src/Chemistry.hpp src/PopIII/popiii.cpp src/PopIII/popiii.hpp src/RadhydroSimulation.hpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp --export-fixes=clang_tidy_review.yaml
164538 warnings generated.
346002 warnings generated.
510727 warnings generated.
677970 warnings generated.
804208 warnings generated.
930446 warnings generated.
1056684 warnings generated.
1182922 warnings generated.
warning: invalid configuration value '0' for option 'modernize-make-shared.IncludeStyle' [clang-tidy-config]
warning: invalid configuration value '0' for option 'modernize-make-unique.IncludeStyle' [clang-tidy-config]
/g/data1b/jh2/ps3459/quokka/src/Chemistry.hpp:37:2: warning: variable 'tiny_profiler_165' of type 'amrex::TinyProfiler' can be declared 'const' [misc-const-correctness]
        BL_PROFILE("Chemistry::computeChemistry()");
        ^
/g/data1b/jh2/ps3459/quokka/extern/amrex/Src/Base/AMReX_BLProfiler.H:487:27: note: expanded from macro 'BL_PROFILE'
#define BL_PROFILE(fname) BL_PROFILE_IMPL(fname, __COUNTER__)
                          ^
/g/data1b/jh2/ps3459/quokka/extern/amrex/Src/Base/AMReX_BLProfiler.H:488:43: note: expanded from macro 'BL_PROFILE_IMPL'
#define BL_PROFILE_IMPL(funame, counter)  amrex::TinyProfiler BL_PROFILE_PASTE(tiny_profiler_, counter)((funame)); \
                                          ^
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:338:20: warning: Value stored to 'x0' during its initialization is never read [clang-analyzer-deadcode.DeadStores]
        amrex::Real const x0 = prob_lo[0] + 0.5 * (prob_hi[0] - prob_lo[0]);
                          ^~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:338:20: note: Value stored to 'x0' during its initialization is never read
        amrex::Real const x0 = prob_lo[0] + 0.5 * (prob_hi[0] - prob_lo[0]);
                          ^~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:339:20: warning: Value stored to 'y0' during its initialization is never read [clang-analyzer-deadcode.DeadStores]
        amrex::Real const y0 = prob_lo[1] + 0.5 * (prob_hi[1] - prob_lo[1]);
                          ^~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:339:20: note: Value stored to 'y0' during its initialization is never read
        amrex::Real const y0 = prob_lo[1] + 0.5 * (prob_hi[1] - prob_lo[1]);
                          ^~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:340:20: warning: Value stored to 'z0' during its initialization is never read [clang-analyzer-deadcode.DeadStores]
        amrex::Real const z0 = prob_lo[2] + 0.5 * (prob_hi[2] - prob_lo[2]);
                          ^~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:340:20: note: Value stored to 'z0' during its initialization is never read
        amrex::Real const z0 = prob_lo[2] + 0.5 * (prob_hi[2] - prob_lo[2]);
                          ^~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:357:4: warning: variable 'cs' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
                        ^
                                    const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:377:4: warning: variable 'Eint' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real Eint = state[bx](i, j, k, HydroSystem<PopIII>::internalEnergy_index);
                        ^
                                    const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:392:4: warning: variable 'Pgas' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real Pgas = HydroSystem<PopIII>::ComputePressure(state[bx], i, j, k);
                        ^
                                    const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:418:4: warning: variable 'pressure' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        Real pressure = HydroSystem<PopIII>::ComputePressure(state[bx], i, j, k);
                        ^
                             const
/g/data1b/jh2/ps3459/quokka/src/PopIII/popiii.cpp:421:4: warning: variable 'cs' of type 'amrex::Real' (aka 'double') can be declared 'const' [misc-const-correctness]
                        amrex::Real cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
                        ^
                                    const
Suppressed 1183244 warnings (1182839 in non-user code, 53 due to line filter, 352 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
```

BenWibking commented:
Can you check what version of clang-tidy is on Gadi with `clang-tidy --version`? Can you do the same for CMake?

Then the next step is to see if you can reproduce the problem with identical versions. Then we can figure out which version is causing the issue.

BenWibking commented:
I also just learned about this, which may help to debug the problem by being able to run the clang-tidy-review action on your laptop: https://github.com/nektos/act

psharda commented:
> Can you check what version of clang-tidy is on Gadi with `clang-tidy --version`? Can you do the same for CMake?
> 

```
ps3459@gadi-login-06:/g/data1b/jh2/ps3459/quokka>clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 15.0.7
  Optimized build.
  Default target: x86_64-redhat-linux-gnu
  Host CPU: cascadelake
```

But the github workflow uses version 14.0?

BenWibking commented:
> But the github workflow uses version 14.0?

Where do you see that? I see `clang-tidy-16`.

BenWibking commented:
Can you re-run with clang-tidy 16?

psharda commented:
> Can you re-run with clang-tidy 16?

Does gadi have clang-tidy 16?

BenWibking commented:
> > Can you re-run with clang-tidy 16?
> 
> Does gadi have clang-tidy 16?

I don't know. You could install it on a macOS laptop with: `brew install llvm@16`. Or you could install it in an Ubuntu virtual machine.

psharda commented:
okay, using clang-tidy v.16 on the laptop gives

```
(base) RSAA-043527:quokka psharda$ clang-tidy -p=build --config-file="src/.clang-tidy" -line-filter='[{"name":"src/Chemistry.hpp","lines":[[34,34],[37,37],[68,72],[86,89],[95,98]]},{"name":"src/PopIII/popiii.cpp","lines":[[1,475]]},{"name":"src/PopIII/popiii.hpp","lines":[[1,23]]},{"name":"src/RadhydroSimulation.hpp","lines":[[116,117],[376,377],[503,503]]}]' src/Chemistry.hpp src/PopIII/popiii.cpp src/PopIII/popiii.hpp src/RadhydroSimulation.hpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp src/TurbDataReader.cpp --export-fixes=clang_tidy_review.yaml
105405 warnings and 1 error generated.
Error while processing /Users/psharda/Documents/quokka/src/Chemistry.hpp.
220750 warnings and 2 errors generated.
Error while processing /Users/psharda/Documents/quokka/src/PopIII/popiii.cpp.
326157 warnings and 3 errors generated.
Error while processing /Users/psharda/Documents/quokka/src/PopIII/popiii.hpp.
434021 warnings and 4 errors generated.
Error while processing /Users/psharda/Documents/quokka/src/RadhydroSimulation.hpp.
526573 warnings and 5 errors generated.
Error while processing /Users/psharda/Documents/quokka/src/TurbDataReader.cpp.
619125 warnings and 6 errors generated.
Error while processing /Users/psharda/Documents/quokka/src/TurbDataReader.cpp.
711677 warnings and 7 errors generated.
Error while processing /Users/psharda/Documents/quokka/src/TurbDataReader.cpp.
804229 warnings and 8 errors generated.
Error while processing /Users/psharda/Documents/quokka/src/TurbDataReader.cpp.
warning: no such sysroot directory: '/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk' [clang-diagnostic-missing-sysroot]
warning: invalid configuration value '0' for option 'modernize-make-shared.IncludeStyle' [clang-tidy-config]
warning: invalid configuration value '0' for option 'modernize-make-unique.IncludeStyle' [clang-tidy-config]
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/math.h:300:15: error: 'math.h' file not found [clang-diagnostic-error]
#include_next <math.h>
              ^~~~~~~~
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/stdlib.h:93:15: error: 'stdlib.h' file not found [clang-diagnostic-error]
#include_next <stdlib.h>
              ^~~~~~~~~~
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/wchar.h:123:15: error: 'wchar.h' file not found [clang-diagnostic-error]
#include_next <wchar.h>
              ^~~~~~~~~
Suppressed 804491 warnings (804126 in non-user code, 79 due to line filter, 286 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Found compiler error(s).
```

BenWibking commented:
Well, a new error that doesn't appear on any other system... I guess that's not helpful.

Can you try running the clang-tidy-review action on your laptop using Act: https://github.com/nektos/act#how-does-it-work? That should ensure it runs with an identical everything compared with what's running on GitHub itself.

psharda committed:
Merge branch 'quokka-astro:development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
updated microp

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
Merge branch 'development' into popiii

psharda committed:
updated

psharda committed:
fix bug

psharda committed:
fix bug

psharda committed:
Merge branch 'development' into popiii

Commit:
dummy commit

Commit:
Merge remote-tracking branch 'origin/test' into popiii

Commit:
debug 1

Commit:
debug 1

Commit:
include other headers present in primordial chem

Commit:
include openPMRsources as in primordial chem

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
test radhydrosim

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
debug chemistry

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
const tiny profiler

Commit:
moved turb files to src

Commit:
updated cmakelists

Commit:
comment print statement

Commit:
codeQL suggestion

Commit:
moved turbdatareader

Commit:
clang-tidy warnings

psharda committed:
Merge branch 'development' into test

Commit:
resolve conflict

Commit:
revert

Commit:
reorder includes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
debug

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
turn off pre-commit

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
remove precommit

Commit:
Merge branch 'popiii' of https://github.com/psharda/quokka into popiii

Commit:
debug

Commit:
dont build primordial chem

psharda committed:
Merge branch 'development' into popiii

Commit:
revert

Commit:
cmake in 3D

Commit:
fix clang-tidy warnings

Commit:
readded precommit

psharda commented:
/azp run

psharda commented on .github/workflows/clang-tidy.yml:
```diff
@@ -23,7 +23,7 @@ jobs:
         apt_packages: libopenmpi-dev,libhdf5-mpi-dev,python3-dev,python3-numpy,python3-matplotlib
-        cmake_command: cmake . -B build -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DQUOKKA_PYTHON=ON -DQUOKKA_OPENPMD=ON -DopenPMD_USE_ADIOS2=OFF
+        cmake_command: cmake . -B build -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DQUOKKA_PYTHON=ON -DQUOKKA_OPENPMD=ON -DopenPMD_USE_ADIOS2=OFF -DAMReX_SPACEDIM=3
         split_workflow: true
```
@BenWibking clang-tidy will pass now. The main issue was that clang-tidy built cmake in 1D, and would review popiii.cpp. But the PopIII test is only built in 3D so that'd why clang-tidy could not find some of the header files. 

However, we perhaps do not want clang-tidy to always run in 3D for every PR. So, what's the long term solution?

BenWibking replied:
Nice debugging!

The goal of all of the testing is to make sure there aren't any bugs in production simulations, so I think running clang-tidy with `AMReX_SPACEDIM=3` is fine. The only real downside is that it won't catch missing `AMREX_D_DECL` macros for indices that might cause 1D tests to fail.

psharda commented:
The 3D GPU tests fail because of errors in `TurbDataReader.cpp`. Could you take a look? It cannot find `amrex::Table3D` and `H5Dread` for some reason. Perhaps amrex Table3D was modified since turbdatareader was written, so the number of arguments are incorrect

BenWibking commented on src/TurbDataReader.cpp:
```diff
@@ -33,14 +33,14 @@ auto read_dataset(hid_t &file_id, char const *dataset_name) -> amrex::Table3D<do
 
 	// allocate array for dataset storage
-	auto *temp_data = new double[data_size];
+	std::unique_ptr<double[]> temp_data(new double[data_size]);
```
This line causes the problem. Why did you change this? It will not work this way because when this function ends, the array will be deallocated, and any attempts to access the contents of `table` will cause a segmentation fault.

BenWibking replied:
`std::unique_ptr` deallocates the memory when it goes out of scope: https://en.cppreference.com/w/cpp/memory/unique_ptr

`amrex::Table3D` is just a wrapper around a raw pointer, so it can't keep it in scope: https://amrex-codes.github.io/amrex/doxygen/structamrex_1_1Table3D.html

BenWibking replied:
I realize clang-tidy is probably not happy with this line. However, as long as we *never* deallocate the memory allocated here, it does not cause a memory safety problem. It is not possible for clang-tidy to check this automatically, so it complains about allocating memory with a raw pointer like this.

The difficulty in fixing this in the way clang-tidy wants is that since `amrex::Table3D` does not allocate its own memory, we have to either i) allocate the memory before calling this function by creating, e.g., a `std::vector`, and then passing a reference to it to this function, or ii) return both an object that allocates its own memory (like `std::vector`, for example) *and* the `amrex::Table3D` object.

Commit:
trying to fix table3d error

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
revert

psharda commented:
/azp run

psharda commented:
all tests should pass now

BenWibking commented on src/CMakeLists.txt:
```diff
@@ -132,6 +132,12 @@ set (QuokkaObjSources "${CMAKE_CURRENT_SOURCE_DIR}/main.cpp" "${CMAKE_CURRENT_SO
+
+
+add_subdirectory(StarCluster)
+#add_subdirectory(PrimordialChem)
```
Is there a reason these need to be listed first?

psharda replied:
Nah, I put them first so that I didn't have to wait for other tests to finish to see if these tests would pass.

BenWibking commented on src/Chemistry.hpp:
```diff
@@ -77,11 +83,19 @@ template <typename problem_t> void computeChemistry(amrex::MultiFab &mf, const R
 			eos(eos_input_re, chemstate);
 
+			if (dt < 0) {
+				amrex::Abort("Cannot do chemistry with dt < 0!");
```
Can you move this check outside of the inner loop? This has severe performance implications on GPU: https://quokka-astro.github.io/quokka/error_checking.html#abort and https://quokka-astro.github.io/quokka/performance.html#gpu-hardware-characteristics

BenWibking commented on src/PopIII/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+        setup_target_for_cuda_compilation(popiii)
+    endif()
+
+    #need h5py to run perturbations.py file below
```
Can you add a comment that explains the purpose of this code is to check whether the h5py python package is installed?

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,479 @@
+		userData_.dv_rms_generated = computeRms(pinned_dvx, pinned_dvy, pinned_dvz);
+		amrex::Print() << "rms dv = " << userData_.dv_rms_generated << "\n";
+
+		const Real rms_dv_target = 1.8050e5;
```
Could this be made a runtime parameter?

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,479 @@
+
+		} else {
+			state.rho = 0.01 * rhotot;
+			state.p = 3.595730e-10; // pressure equilibrium - this is the pressure within the core
```
How did you compute this value?

psharda replied:
Ah, I pre-calculated it during testing and debugging to make life easier but never updated the code to automatically calculate it. Done now.

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,479 @@
+			amrex::Real const cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
+
+			const amrex::Real l_Jeans = cs * std::sqrt(M_PI / (G * rho));
+			if (l_Jeans < (N_cells * dx) && rho > 2e-20) {
```
Could the density threshold be a parameter instead of hard-coded to `2e-20` here?

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,479 @@
+
+		amrex::ParallelFor(mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+			Real const rho = state[bx](i, j, k, HydroSystem<PopIII>::density_index);
+			amrex::Real const Eint = state[bx](i, j, k, HydroSystem<PopIII>::internalEnergy_index);
```
You should compute this as Etot - KE instead.

psharda replied:
Why?

BenWibking replied:
You're right, it should be fine, since they are sync'd after each timestep.

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,479 @@
+		});
+	}
+
+	if (dname == "velx") {
```
This is fine, but it might make sense to add vely, velz.

psharda replied:
Yes, but since this is just meant to be a prototype for a production simulation it is fine to show an example of a derived var.

BenWibking commented on src/PopIII/popiii.cpp:
```diff
@@ -0,0 +1,479 @@
+	RadhydroSimulation<PopIII> sim(BCs_cc);
+	sim.doPoissonSolve_ = 1; // enable self-gravity
+	sim.densityFloor_ = 1e-25;
+	sim.tempFloor_ = 2.73 * (30.0 + 1.0);
```
These values can go in the *.in file.

psharda replied:
All three of these? How do you define them in the .in file? Do you just copy and paste them as it is, with the `sim.` prefix? And do you need parmparse to read them in?

BenWibking replied:
Just the density floor and temperature floor. It currently can't parse expressions like that for the temperature floor, so you could leave that as is. I don't think there's one for the Poisson solve either. They're in this table: https://quokka-astro.github.io/quokka/parameters.html#general

BenWibking replied:
@psharda You don't need to do it again, it's already ParmParsed in simulation.hpp: https://github.com/quokka-astro/quokka/blob/ae42169d912724ee3642b79ac1d493964a87e801/src/simulation.hpp#L608

psharda replied:
@BenWibking so I can completely remove the line `sim.densityFloor_ ` from popiii.cpp?

psharda replied:
Can you see if the latest commit is fine?

BenWibking commented on tests/PopIII.in:
```diff
@@ -0,0 +1,100 @@
+amr.n_cell          = 64 64 64
+amr.max_level       = 0     # number of levels = max_level + 1
+amr.blocking_factor = 32    # grid size must be divisible by this
+amr.max_grid_size   = 1024    # at least 128 for GPUs
```
This is too big. It should be 128 or 256 at most.

BenWibking reviewed (changes requested):
A few minor comments.

Commit:
arrange tests alphabetically

Commit:
reduce grid size to 128

Commit:
comment for h5py

Commit:
make initial rms vel a runtime param

Commit:
let the code calculate pressure

Commit:
make jeans refinement params runtime params

Commit:
move dt < 0 outside of the inner loop

psharda commented:
/azp run

Commit:
declaring an initial value of vars

Commit:
make density floor runtime param

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
use a different variable for parmparse

psharda commented:
/azp run

Commit:
dont need to define density_floor, its defined in simulation.hpp

BenWibking commented:
/azp run
